AK/SipHash.cpp

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) 2023, kleines Filmröllchen <filmroellchen@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/MemoryStream.h>
+#include <AK/SipHash.h>
+#include <AK/Span.h>
+#include <AK/UFixedBigInt.h>
+
+#ifdef KERNEL
+#    include <Kernel/Security/Random.h>
+#else
+#    include <AK/Random.h>
+#endif
+
+namespace AK {
+
+#define ROTL(x, b) (u64)(((x) << (b)) | ((x) >> (64 - (b))))
+
+#define SIPROUND           \
+    do {                   \
+        v0 += v1;          \
+        v1 = ROTL(v1, 13); \
+        v1 ^= v0;          \
+        v0 = ROTL(v0, 32); \
+        v2 += v3;          \
+        v3 = ROTL(v3, 16); \
+        v3 ^= v2;          \
+        v0 += v3;          \
+        v3 = ROTL(v3, 21); \
+        v3 ^= v0;          \
+        v2 += v1;          \
+        v1 = ROTL(v1, 17); \
+        v1 ^= v2;          \
+        v2 = ROTL(v2, 32); \
+    } while (0)
```

Comment:
Let's circle back to the implementation now; please utilise inline (constexpr) functions instead of this C-style nonsense.